tests/test_connection.py

```diff
@@ -0,0 +1,189 @@
+    @mock.patch('syncano.connection.Connection.make_request')
+    def test_request_authentication(self, make_request_mock, authenticate_mock):
+        self.assertFalse(make_request_mock.called)
+        self.assertFalse(authenticate_mock.called)
+        self.connection.request('POST', 'test')
+        self.assertTrue(make_request_mock.called)
+        self.assertTrue(authenticate_mock.called)
+
+    @mock.patch('syncano.connection.Connection.authenticate')
+    @mock.patch('syncano.connection.Connection.make_request')
+    def test_request_serialization(self, make_request_mock, authenticate_mock):
+        self.assertFalse(make_request_mock.called)
+        self.assertFalse(authenticate_mock.called)
+
+        make_request_mock.return_value = {'a': 1}
+        content = self.connection.request('POST', 'test')
+
+        self.assertTrue(make_request_mock.called)
+        self.assertTrue(authenticate_mock.called)
+        self.assertEqual(content, make_request_mock.return_value)
+
+        content = self.connection.request('POST', 'test', result_class=mock.MagicMock)
+        self.assertIsInstance(content, mock.MagicMock)
+
+        make_request_mock.return_value = {
+            'objects': [],
+            'next': None,
+            'prev': None,
+        }
+        content = self.connection.request('POST', 'test')
+        self.assertIsInstance(content, self.connection.RESULT_SET_CLASS)
+
+    @mock.patch('requests.Session.post')
+    def test_make_request(self, post_mock):
+        response_mock = mock.MagicMock(
+            status_code=200,
+            headers={
+                'content-type': 'application/json'
+            },
+        )
+        response_mock.json.return_value = {'ok': 'ok'}
+        post_mock.return_value = response_mock
+
+        out = self.connection.make_request('POST', 'test')
+        self.assertTrue(post_mock.called)
+        self.assertEqual(out, response_mock.json.return_value)
+
+    def test_invalid_method_name(self):
+        with self.assertRaises(SyncanoValueError):
+            self.connection.make_request('INVALID', 'test')
+
+    @mock.patch('requests.Session.post')
+    def test_request_error(self, post_mock):
+        post_mock.return_value = mock.MagicMock(status_code=404, text='Invalid request')
+        self.assertFalse(post_mock.called)
+
+        with self.assertRaises(SyncanoRequestError):
+            self.connection.make_request('POST', 'test')
+
+        self.assertTrue(post_mock.called)
+
+    def test_is_authenticated(self):
+        self.assertFalse(self.connection.is_authenticated())
+        self.connection.api_key = 'xxxx'
+        self.assertTrue(self.connection.is_authenticated())
+
+    @mock.patch('syncano.connection.Connection.make_request')
+    def test_already_authenticated(self, make_request_mock):
+        self.assertIsNone(self.connection.api_key)
+        self.assertFalse(make_request_mock.called)
+
+        self.connection.api_key = 'Ala has cat'
```

Comment:
has a cat ;) or the cat. I really like strange testing strings ;), they reveal a lot about developer's current mental state.